Mapper with implementationPackage on abstract class not working with componentModel=default

I have 2 mappers in the same package. Both annotated with `@Mapper (implementationPackage = "...")`.

When I try to generate the mappers I get the exception:
```
[INFO] --- maven-processor-plugin:2.2.4:process (process) @ web ---
[ERROR] diagnostic: /Users/terciofilho/Workspace/master/web/src/main/java/br/com/web/controller/mapper/AnuncioMapper.java:29: error: Internal error in the mapping processor: java.lang.RuntimeException: javax.annotation.processing.FilerException: Attempt to reopen a file for path /Users/terciofilho/Workspace/master/web/target/classes/META-INF/services/br.com.web.controller.mapper.null  
	at org.mapstruct.ap.internal.processor.MapperServiceProcessor.createSourceFile(MapperServiceProcessor.java:92)  
	at org.mapstruct.ap.internal.processor.MapperServiceProcessor.writeToSourceFile(MapperServiceProcessor.java:76)  
	at org.mapstruct.ap.internal.processor.MapperServiceProcessor.process(MapperServiceProcessor.java:62)  
	at org.mapstruct.ap.internal.processor.MapperServiceProcessor.process(MapperServiceProcessor.java:43)  
	at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:223)  
	at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:203)  
	at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:169)  
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)  
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)  
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)  
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)  
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)  
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)  
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)  
	at com.sun.tools.javac.main.Main.compile(Main.java:523)  
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)  
	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)  
	at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.executeWithExceptionsHandled(AbstractAnnotationProcessorMojo.java:545)  
	at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.execute(AbstractAnnotationProcessorMojo.java:285)  
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)  
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)  
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)  
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)  
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)  
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)  
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)  
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)  
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)  
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)  
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)  
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)  
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)  
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)  
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)  
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)  
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)  
	at java.lang.reflect.Method.invoke(Method.java:497)  
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)  
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)  
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)  
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: javax.annotation.processing.FilerException: Attempt to reopen a file for path /Users/terciofilho/Workspace/master/web/target/classes/META-INF/services/br.com.web.controller.mapper.null
	at com.sun.tools.javac.processing.JavacFiler.checkFileReopening(JavacFiler.java:535)  
	at com.sun.tools.javac.processing.JavacFiler.createResource(JavacFiler.java:431)  
	at org.mapstruct.ap.internal.processor.MapperServiceProcessor.createSourceFile(MapperServiceProcessor.java:89)  	... 40 more
public abstract class AnuncioMapper extends BasicMapper {
                ^
[ERROR] error on execute: error during compilation
```

If I remove the `implementationPackage` property of @Mapper, no exception is thrown and the mapper is generated correctly.

--- a/processor/src/main/java/org/mapstruct/ap/internal/processor/MapperServiceProcessor.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/processor/MapperServiceProcessor.java
@@ -18,18 +18,19 @@
  */
 package org.mapstruct.ap.internal.processor;
 
+import java.io.IOException;
+
+import javax.annotation.processing.Filer;
+import javax.lang.model.element.TypeElement;
+import javax.tools.FileObject;
+import javax.tools.StandardLocation;
+
 import org.mapstruct.ap.internal.model.GeneratedType;
 import org.mapstruct.ap.internal.model.Mapper;
 import org.mapstruct.ap.internal.model.ServicesEntry;
 import org.mapstruct.ap.internal.option.OptionsHelper;
 import org.mapstruct.ap.internal.util.MapperConfiguration;
 import org.mapstruct.ap.internal.writer.ModelWriter;
-
-import javax.annotation.processing.Filer;
-import javax.lang.model.element.TypeElement;
-import javax.tools.FileObject;
-import javax.tools.StandardLocation;
-import java.io.IOException;
 
 /**
  * A {@link ModelElementProcessor} which creates files in the {@code META-INF/services}
@@ -77,7 +78,9 @@ public class MapperServiceProcessor  implements ModelElementProcessor<Mapper, Vo
     }
 
     private ServicesEntry getServicesEntry(GeneratedType model) {
-        return new ServicesEntry(model.getInterfacePackage(), model.getInterfaceName(),
+        String mapperName = model.getInterfaceName() != null ? model.getInterfaceName() : model.getSuperClassName();
+
+        return new ServicesEntry(model.getInterfacePackage(), mapperName,
                                  model.getPackageName(), model.getName());
     }
 

--- a/processor/src/test/java/org/mapstruct/ap/test/destination/AbstractDestinationClassNameMapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/destination/AbstractDestinationClassNameMapper.java
@@ -1,0 +1,33 @@
+/**
+ *  Copyright 2012-2015 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.destination;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Andreas Gudian
+ */
+@Mapper(implementationName = "My<CLASS_NAME>CustomImpl")
+public abstract class AbstractDestinationClassNameMapper {
+    public static final AbstractDestinationClassNameMapper INSTANCE =
+        Mappers.getMapper( AbstractDestinationClassNameMapper.class );
+
+    public abstract String intToString(Integer source);
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/destination/AbstractDestinationPackageNameMapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/destination/AbstractDestinationPackageNameMapper.java
@@ -1,0 +1,33 @@
+/**
+ *  Copyright 2012-2015 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.destination;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Andreas Gudian
+ */
+@Mapper(implementationPackage = "<PACKAGE_NAME>.dest")
+public abstract class AbstractDestinationPackageNameMapper {
+    public static final AbstractDestinationPackageNameMapper INSTANCE =
+        Mappers.getMapper( AbstractDestinationPackageNameMapper.class );
+
+    public abstract String intToString(Integer source);
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/destination/DestinationClassNameTest.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/destination/DestinationClassNameTest.java
@@ -85,4 +85,18 @@ public class DestinationClassNameTest {
         assertThat( ( (DestinationClassNameMapperDecorator) instance ).delegate.getClass().getSimpleName() )
                 .isEqualTo( "MyDestinationClassNameMapperDecoratedCustomImpl_" );
     }
+
+    @Test
+    @WithClasses({ AbstractDestinationClassNameMapper.class, AbstractDestinationPackageNameMapper.class })
+    public void shouldWorkWithAbstractClasses() {
+        AbstractDestinationClassNameMapper mapper1 = AbstractDestinationClassNameMapper.INSTANCE;
+        assertThat( mapper1.getClass().getPackage().getName() )
+                .isEqualTo( AbstractDestinationClassNameMapper.class.getPackage().getName() );
+        assertThat( mapper1.getClass().getSimpleName() ).isEqualTo( "MyAbstractDestinationClassNameMapperCustomImpl" );
+
+        AbstractDestinationPackageNameMapper mapper2 = AbstractDestinationPackageNameMapper.INSTANCE;
+        assertThat( mapper2.getClass().getPackage().getName() )
+                .isEqualTo( AbstractDestinationPackageNameMapper.class.getPackage().getName() + ".dest" );
+        assertThat( mapper2.getClass().getSimpleName() ).isEqualTo( "AbstractDestinationPackageNameMapperImpl" );
+    }
 }

